charmcraft/extensions/extension.py

```diff
@@ -0,0 +1,143 @@
+# Copyright 2018-2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Extension base class definition."""
+
+import abc
+import os
+import sys
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Sequence, Tuple, final
+
+from craft_cli import emit
+
+from charmcraft import errors
+
+
+class Extension(abc.ABC):
+    """Extension is the class from which all extensions inherit.
+
+    Extensions have the ability to add snippets to apps, parts, and indeed add new parts
+    to a given charmcraft.yaml.
+    :ivar project_root: the root of the project.
+    :ivar yaml_data: the raw yaml data.
+    """
+
+    def __init__(
+        self,
+        *,
+        project_root: Path,
+        yaml_data: Dict[str, Any],
+    ) -> None:
+        """Create a new Extension."""
+        self.project_root = project_root
+        self.yaml_data = yaml_data
+
+    @staticmethod
+    @abc.abstractmethod
+    def get_supported_bases() -> List[Tuple[str, ...]]:
```

Comment:
What about

```suggestion
    def get_supported_bases() -> Tuple[charmcraft.bases.Base, ...]:
```

or possibly a `Tuple[Dict[str, str], ...]`

Reply:
The base is not exist at that moment.

Reply:
Then it's probably best with the dict option, since our charmcraft bases naturally contain a `name` and a `channel`

Reply:
Ahh ok. I misread this, never mind :-)